example/service/s3/listObjects/listObjects.go

```diff
@@ -0,0 +1,129 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"log"
+
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+)
+
+var (
+	bucketName      string
+	objectPrefix    string
+	objectDelimiter string
+)
+
+func init() {
+	flag.StringVar(&bucketName, "bucket", "", "The `name` of the S3 bucket to list objects from.")
+	flag.StringVar(&objectPrefix, "prefix", "", "The optional `object prefix` of the S3 Object keys to list.")
+	flag.StringVar(&objectDelimiter, "delimiter", "/",
+		"The optional `object key delimiter` used by S3 List objects to group object keys.")
+}
+
+// Lists all objects in a bucket using pagination
+//
+// Usage:
+// listObjects <bucket>
+func main() {
+	flag.Parse()
+	if len(bucketName) == 0 {
+		flag.PrintDefaults()
+		log.Fatalf("invalid parameters, bucket name required")
+	}
+
+	// Load the SDK's configuration from environment and shared config, and
+	// create the client with this.
+	cfg, err := config.LoadDefaultConfig()
+	if err != nil {
+		log.Fatalf("failed to load SDK configuration, %v", err)
+	}
+
+	client := s3.NewFromConfig(cfg)
+
+	// Set the parameters based on the CLI flag inputs.
+	params := &s3.ListObjectsV2Input{
+		Bucket: &bucketName,
+	}
+	if len(objectPrefix) != 0 {
+		params.Prefix = &objectPrefix
+	}
+	if len(objectDelimiter) != 0 {
+		params.Delimiter = &objectDelimiter
+	}
+
+	// TODO replace this with the code generate paginator when available
+	// s3.NewListObjectsV2Paginator()
+	p := NewS3ListObjectsV2Paginator(client, params)
+
+	// Iterate through the S3 object pages, printing each object returned.
+	var i int
+	log.Println("Objects:")
+	for p.HasMorePages() {
+		i++
+
+		page, err := p.NextPage(context.TODO())
+		if err != nil {
+			log.Fatalf("failed to get page %v, %v", i, err)
+		}
+
+		// Log the objects found
+		for _, obj := range page.Contents {
+			fmt.Println("Object:", *obj.Key)
+		}
+	}
+}
+
+// S3ListObjectsV2APIClient provides interface for the S3 API client
+// ListObjectsV2 operation call.
+type S3ListObjectsV2APIClient interface {
+	ListObjectsV2(context.Context, *s3.ListObjectsV2Input, ...func(*s3.Options)) (*s3.ListObjectsV2Output, error)
+}
+
+// S3ListObjectsV2Paginator provides the paginator to paginate S3 ListObjectsV2
+// response pages.
+type S3ListObjectsV2Paginator struct {
+	client S3ListObjectsV2APIClient
+	params *s3.ListObjectsV2Input
+
+	nextToken *string
+	firstPage bool
+}
+
+// NewS3ListObjectsV2Paginator initializes a new S3 ListObjectsV2 Paginator for
+// paginating the ListObjectsV2 respones.
+func NewS3ListObjectsV2Paginator(client S3ListObjectsV2APIClient, params *s3.ListObjectsV2Input) *S3ListObjectsV2Paginator {
```

Comment:
This should probably have a page size parameter

Reply:
Page size is currently a parameter of the `s3.ListObjectsV2Input` type. I'm not sure if this should be duplicated here or not since it is an optional parameter. I'd think if the user wants to specify this value with a custom size then it should be done on the input value.

What are your thoughts on this?

Reply:
The reason it's an attribute on the paginator trait is so that clients can expose a consistent interface for providing that value. Between operations and services the member could be any number of things, like `Limit`, `MaxItems`, `PageSize`, etc.

Reply:
yup though its still an optional parameter in all APIs. By putting it as a parameter on the constructor the SDK is paginator is effectively making it required. The handling for required, but not required parameter, would be to treat the limit parameter as unset if its value is 0.  Minor, but this approach still requires the user to pass in a `0` for the default behavior.

Reply:
Well you can always make a pointer so people pass it in as nil, or have a function to set it as a second step that takes a non-pointer

Reply:
Adding functional options to add configuration to the paginator might be a good way to do this.

```go
type ListObjectsV2PaginatorOptions struct {
     Limit int32 // generic name for all paginators
}

func NewListObjectsV2Paginator(client ListObjectsV2APIClient, params *s3.ListObjectsV2Input,
     optFns ...func(*ListObjectsV2PaginatorOptions)
```

```go
p := s3.NewListObjectsV2Paginator(client, params, func(o *ListObjectsV2PaginatorOptions) {
      o.Limit = 10
})
```

Reply:
updated to include this pattern 👍